client/trino-client/src/main/java/io/trino/client/StatementStats.java

```diff
@@ -194,12 +206,56 @@ public StageStats getRootStage()
     @JsonProperty
     public OptionalDouble getProgressPercentage()
     {
+        if (faultTolerantExecution) {
+            if (rootStage == null) {
+                return OptionalDouble.empty();
+            }
+            // Unlike pipelined execution, fault tolerant execution doesn't execute stages all at
+            // once and some stages will be in PLANNED state in the middle of execution. Therefore,
+            // we don't know the number of total splits upfront.
+            //
+            // To avoid the confusion of progress bar going back and forth, we calculate the
+            // progress percentage of each stage (0 if not scheduled), and average them.
+            double percentageSum = 0.0;
+            int totalStages = 0;
+            Queue<StageStats> queue = new ArrayDeque<>();
+            queue.add(rootStage);
+            while (!queue.isEmpty()) {
+                StageStats stage = queue.poll();
+                totalStages++;
+                if (stage.isScheduled()) {
+                    percentageSum += 100.0 * stage.getCompletedSplits() / stage.getTotalSplits();
+                }
+                queue.addAll(stage.getSubStages());
+            }
+            return OptionalDouble.of(min(100, percentageSum / totalStages));
+        }
+
         if (!scheduled || totalSplits == 0) {
             return OptionalDouble.empty();
         }
         return OptionalDouble.of(min(100, (completedSplits * 100.0) / totalSplits));
     }
 
+    public int getFaultTolerantExecutionRunningPercentage()
```

Comment:
nit: maybe extract the implementation into a helper method (you can pass a boolean flag on what to compute)